### PR TITLE
Rename template args of Point and Tensor

### PR DESCRIPTION
**Pull request by bangerth**
Since we decided to shelve #526 , this patch updates the name of the type argument of `Point` and `Tensor` from the generic `Number` to the more specific name `CoordinateType`. The intent is to better document what this type is used for. The commits in this PR also add documentation for these template arguments.

### CODE REVIEW

**Comment by bangerth** [comment]
Or would `ElementType` be a better name for the elements of tensors (and `CoordinateType` for `Point`)?

**Comment by tjhei** [comment]
What is the reason to get rid of ```Number```? It is used in several places of the library (```SparseMatrix```, ```Vector```, ...) and always has the same meaning and default.

**Comment by bangerth** [comment]
I realized when thinking about `Function` how just using a generic name for the type is not really conducive to understanding what it is used for. `Function` was converted with the template argument because the names of the template arguments didn't give away their meaning. I suppose this wouldn't have happened if we had chosen more descriptive names. I really believe that choosing good names helps understand *intent*.

**Comment by bangerth** [comment]
In other words, I would approve pull requests that change `number` into `ElementType` or `ScalarType` etc for `Vector`, `SparseMatrix`, etc.

**Comment by tjhei** [comment]
Okay, let me rephrase:
- I think Vector, SparseMatrix, Tensor, Point, etc. should have the same name for their element/entry data type. This makes it obvious for the user to figure out what to plug in (similar to ```dim```).
- I don't feel strongly about the name itself: number, ScalarType, or ElementType is fine (CoordinateType is not so great). Note that we don't use "Type" as part of template parameters in other places.
- There is some value in keeping the name we currently have (familiarity). Again, this is just my opinion. Maybe others can put in their vote.

**Comment by tamiko** [comment]
I pretty much agree with all remarks pointed out by Timo.

>     I think Vector, SparseMatrix, Tensor, Point, etc. should have the
>     same name for their element/entry data type. This makes it
>     obvious for the user to figure out what to plug in (similar to
>     dim).

Yes, please!

>     Note that we don't use "Type" as part of template parameters in
>     other places.

Given the fact that templates are a form of type polymorphism, the
"Type" in the name is pretty much redundant. Further, we already make
clear that it is a type template by using "typename" (instead of
"class").

**Comment by kronbichler** [comment]
Regarding matrices and vectors, Scalar or ScalarType is the most intuitive name in my opinion. And it is definitely good to have the same name throughout the library. In my opinion, this is also more important than putting extra emphasis on the template parameter name in Point. So I agree with @tjhei and @tamiko.

**Comment by bangerth** [comment]
OK, I hear you :-) I've backed out the change in name of the template argument. The patch now only contains a variety of documentation updates that I hope should be uncontroversial.

File: include/deal.II/base/point.h [hunk 1]
```diff
@@ -171,38 +189,47 @@ class Point : public Tensor<1,dim,Number>
...
-   * this does not need to copy a point at least once.
+   * Divide the current point by a factor.
    */
   Point<dim,Number>   operator / (const Number) const;
 
   /**
-   * Returns the scalar product of this point vector with itself, i.e. the
+   * Return the scalar product of the vectors representing two points.
+   */
+  Number              operator * (const Tensor<1,dim,Number> &) const;
```
Comment (tjhei) [inline include/deal.II/base/point.h]: should we name the argument?

File: include/deal.II/base/tensor_base.h [hunk 2]
```diff
@@ -18,7 +18,7 @@
 
 
 // single this file out from tensor.h, since we want to derive
-// Point<dim,Number> from Tensor<1,dim,Number>. However, the point class will
+// Point<dim,CoordinateType> from Tensor<1,dim,Number>. However, the point class will
```
Comment (tjhei) [inline include/deal.II/base/tensor_base.h]: replace ```CoordinateType``` with ```Number```

**Comment by tjhei** [comment]
Looks good and uncontroversial otherwise. :-)

**Comment by bangerth** [comment]
OK, I've addressed @tjhei's two points.

**Comment by tamiko** [comment]
Please squash and merge.
